### ISSUE TITLE
Prevent use of strings with code points over 0xFF as arguments to "vec"

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#17021]
Migrated from [rt.perl.org#134139](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=134139) (status was 'pending release')

Searchable as RT134139$

**Comment by p5pRT** [comment]
### From @jkeenan

pod/perldeprecation\.pod contains this entry for perl\-5\.32&#8203;:

\#\#\#\#\#
Use of strings with code points over 0xFF as arguments to
"vec"

"vec" views its string argument as a sequence of bits\. A
string containing a code point over 0xFF is nonsensical\.
This usage is deprecated in Perl 5\.28\, and will be removed
in Perl 5\.32\.
\#\#\#\#\#

This entry was \(mostly\) added in this commit&#8203;:

\#\#\#\#\#
commit 76aae383784f602979d5efc9c8828918d67827d8
Author&#8203;: Karl Williamson \<khw@&#8203;cpan\.org>
Date&#8203;:   Thu Jun 1 16&#8203;:10&#8203;:48 2017 \-0600

perldeprecation&#8203;: Add vec\(\) with above\-FF code points
\#\#\#\#\#

Make it so\.

**Comment by p5pRT** [comment]
### From @jkeenan

Summary of my perl5 \(revision 5 version 31 subversion 0\) configuration&#8203;:
&nbsp;  Commit id&#8203;: 58f4626762668e1c1948832073998af84b2c85d0
&nbsp;  Platform&#8203;:
&nbsp;    osname=linux
&nbsp;    osvers=4\.15\.0\-50\-generic
&nbsp;    archname=x86\_64\-linux
&nbsp;    uname='linux zareason 4\.15\.0\-50\-generic \#54\-ubuntu smp mon may 6 18&#8203;:46&#8203;:08 utc 2019 x86\_64 x86\_64 x86\_64 gnulinux '
&nbsp;    config\_args='\-des \-Dusedevel \-Dusemymalloc=y'
&nbsp;    hint=recommended
&nbsp;    useposix=true
&nbsp;    d\_sigaction=define
&nbsp;    useithreads=undef
&nbsp;    usemultiplicity=undef
&nbsp;    use64bitint=define
&nbsp;    use64bitall=define
&nbsp;    uselongdouble=undef
&nbsp;    usemymalloc=y
&nbsp;    default\_inc\_excludes\_dot=define
&nbsp;    bincompat5005=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='cc'
&nbsp;    ccflags ='\-fwrapv \-fno\-strict\-aliasing \-pipe \-fstack\-protector\-strong \-I/usr/local/include \-D\_LARGEFILE\_SOURCE \-D\_FILE\_OFFSET\_BITS=64'
&nbsp;    optimize='\-O2'
&nbsp;    cppflags='\-fwrapv \-fno\-strict\-aliasing \-pipe \-fstack\-protector\-strong \-I/usr/local/include'
&nbsp;    ccversion=''
&nbsp;    gccversion='7\.4\.0'
&nbsp;    gccosandvers=''
&nbsp;    intsize=4
&nbsp;    longsize=8
&nbsp;    ptrsize=8
&nbsp;    doublesize=8
&nbsp;    byteorder=12345678
&nbsp;    doublekind=3
&nbsp;    d\_longlong=define
&nbsp;    longlongsize=8
&nbsp;    d\_longdbl=define
&nbsp;    longdblsize=16
&nbsp;    longdblkind=3
&nbsp;    ivtype='long'
&nbsp;    ivsize=8
&nbsp;    nvtype='double'
&nbsp;    nvsize=8
&nbsp;    Off\_t='off\_t'
&nbsp;    lseeksize=8
&nbsp;    alignbytes=8
&nbsp;    prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='cc'
&nbsp;    ldflags =' \-fstack\-protector\-strong \-L/usr/local/lib'
&nbsp;    libpth=/usr/local/lib /usr/lib/gcc/x86\_64\-linux\-gnu/7/include\-fixed /usr/include/x86\_64\-linux\-gnu /usr/lib /lib/x86\_64\-linux\-gnu /lib/\.\./lib /usr/lib/x86\_64\-linux\-gnu /usr/lib/\.\./lib /lib /lib64 /usr/lib64
&nbsp;    libs=\-lpthread \-lnsl \-lgdbm \-ldb \-ldl \-lm \-lcrypt \-lutil \-lc
&nbsp;    perllibs=\-lpthread \-lnsl \-ldl \-lm \-lcrypt \-lutil \-lc
&nbsp;    libc=libc\-2\.27\.so
&nbsp;    so=so
&nbsp;    useshrplib=false
&nbsp;    libperl=libperl\.a
&nbsp;    gnulibc\_version='2\.27'
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_dlopen\.xs
&nbsp;    dlext=so
&nbsp;    d\_dlsymun=undef
&nbsp;    ccdlflags='\-Wl\,\-E'
&nbsp;    cccdlflags='\-fPIC'
&nbsp;    lddlflags='\-shared \-O2 \-L/usr/local/lib \-fstack\-protector\-strong'


Characteristics of this binary \(from libperl\)&#8203;: 
&nbsp;  Compile\-time options&#8203;:
&nbsp;    HAS\_TIMES
&nbsp;    MYMALLOC
&nbsp;    PERLIO\_LAYERS
&nbsp;    PERL\_COPY\_ON\_WRITE
&nbsp;    PERL\_DONT\_CREATE\_GVSV
&nbsp;    PERL\_MALLOC\_WRAP
&nbsp;    PERL\_OP\_PARENT
&nbsp;    PERL\_PRESERVE\_IVUV
&nbsp;    PERL\_USE\_DEVEL
&nbsp;    USE\_64\_BIT\_ALL
&nbsp;    USE\_64\_BIT\_INT
&nbsp;    USE\_LARGE\_FILES
&nbsp;    USE\_LOCALE
&nbsp;    USE\_LOCALE\_COLLATE
&nbsp;    USE\_LOCALE\_CTYPE
&nbsp;    USE\_LOCALE\_NUMERIC
&nbsp;    USE\_LOCALE\_TIME
&nbsp;    USE\_PERLIO
&nbsp;    USE\_PERL\_ATOF
&nbsp;  Built under linux
&nbsp;  Compiled at May 22 2019 14&#8203;:58&#8203;:04
&nbsp;  %ENV&#8203;:
&nbsp;    PERL2DIR="/home/jkeenan/gitwork/perl2"
&nbsp;    PERLBREW\_HOME="/home/jkeenan/\.perlbrew"
&nbsp;    PERLBREW\_MANPATH="/home/jkeenan/perl5/perlbrew/perls/perl\-5\.28\.0/man"
&nbsp;    PERLBREW\_PATH="/home/jkeenan/perl5/perlbrew/bin&#8203;:/home/jkeenan/perl5/perlbrew/perls/perl\-5\.28\.0/bin"
&nbsp;    PERLBREW\_PERL="perl\-5\.28\.0"
&nbsp;    PERLBREW\_ROOT="/home/jkeenan/perl5/perlbrew"
&nbsp;    PERLBREW\_SHELLRC\_VERSION="0\.84"
&nbsp;    PERLBREW\_VERSION="0\.84"
&nbsp;    PERL\_WORKDIR="/home/jkeenan/gitwork/perl"
&nbsp;  @&#8203;INC&#8203;:
&nbsp;    lib
&nbsp;    /usr/local/lib/perl5/site\_perl/5\.31\.0/x86\_64\-linux
&nbsp;    /usr/local/lib/perl5/site\_perl/5\.31\.0
&nbsp;    /usr/local/lib/perl5/5\.31\.0/x86\_64\-linux
&nbsp;    /usr/local/lib/perl5/5\.31\.0

**Comment by p5pRT** [comment]
### From @jkeenan

On Sat\, 25 May 2019 02&#8203;:08&#8203;:59 GMT\, jkeenan@&#8203;pobox\.com wrote&#8203;:
> pod/perldeprecation\.pod contains this entry for perl\-5\.32&#8203;:
> 
> \#\#\#\#\#
> Use of strings with code points over 0xFF as arguments to
> "vec"
> 
> "vec" views its string argument as a sequence of bits\. A
> string containing a code point over 0xFF is nonsensical\.
> This usage is deprecated in Perl 5\.28\, and will be removed
> in Perl 5\.32\.
> \#\#\#\#\#
> 
> This entry was \(mostly\) added in this commit&#8203;:
> 
> \#\#\#\#\#
> commit 76aae383784f602979d5efc9c8828918d67827d8
> Author&#8203;: Karl Williamson \<khw@&#8203;cpan\.org>
> Date&#8203;:   Thu Jun 1 16&#8203;:10&#8203;:48 2017 \-0600
> 
> perldeprecation&#8203;: Add vec\(\) with above\-FF code points
> \#\#\#\#\#
> 
> Make it so\.

Please review the patch attached\.  This is available for smoking in the following branch&#8203;:

smoke\-me/jkeenan/rt\-134139\-vec

Thank you very much\.
\-\- 
James E Keenan \(jkeenan@&#8203;cpan\.org\)

**Comment by p5pRT** [comment]
### From @jkeenan

<details><summary>0001-Use-of-strings-with-code-points-over-0xFF-as-argumen.patch</summary>

```diff
From b137856f094821c8ad330f6a5768544ae59555d0 Mon Sep 17 00:00:00 2001
From: James E Keenan <jkeenan@cpan.org>
Date: Sun, 26 May 2019 22:30:40 -0400
Subject: [PATCH] Use of strings with code points over 0xFF as arguments to
 "vec"

Implement scheduled fatalization.  Adapt existing tests in t/op/vec.t.
Eliminate t/lib/warnings/doop and move one test to t/op/vec.t.

Document this fatalization in perldiag and perlfunc.

For: RT # 134139
---
 MANIFEST            |  1 -
 doop.c              |  5 +----
 pod/perldiag.pod    |  7 +++----
 pod/perlfunc.pod    |  6 ++----
 t/lib/warnings/doop | 14 --------------
 t/op/vec.t          | 26 +++++++++++++++++---------
 6 files changed, 23 insertions(+), 36 deletions(-)
 delete mode 100644 t/lib/warnings/doop

diff --git a/MANIFEST b/MANIFEST
index de891fc8ff..830340475f 100644
--- a/MANIFEST
+++ b/MANIFEST
@@ -5507,7 +5507,6 @@ t/lib/warnings/9enabled		Tests warnings
 t/lib/warnings/9uninit		Tests "Use of uninitialized" warnings
 t/lib/warnings/av		Tests for av.c for warnings.t
 t/lib/warnings/doio		Tests for doio.c for warnings.t
-t/lib/warnings/doop		Tests for doop.c for warnings.t
 t/lib/warnings/gv		Tests for gv.c for warnings.t
 t/lib/warnings/hv		Tests for hv.c for warnings.t
 t/lib/warnings/malloc		Tests for malloc.c for warnings.t
diff --git a/doop.c b/doop.c
index 54e35f10a6..3b4377c652 100644
--- a/doop.c
+++ b/doop.c
@@ -763,10 +763,7 @@ Perl_do_vecget(pTHX_ SV *sv, STRLEN offset, int size)
             s = (unsigned char *) SvPV_flags(sv, srclen, svpv_flags);
         }
         else {
-            Perl_ck_warner_d(aTHX_ packWARN(WARN_DEPRECATED),
-                                "Use of strings with code points over 0xFF as"
-                                " arguments to vec is deprecated. This will"
-                                " be a fatal error in Perl 5.32");
+	        Perl_croak(aTHX_ "Use of strings with code points over 0xFF as arguments to vec is forbidden");
         }
     }
 
diff --git a/pod/perldiag.pod b/pod/perldiag.pod
index f69b1b8367..11339f0e9b 100644
--- a/pod/perldiag.pod
+++ b/pod/perldiag.pod
@@ -7340,13 +7340,12 @@ operators treat their operands as strings of bytes, and values beyond
 
 This became fatal in Perl 5.28.
 
-=item Use of strings with code points over 0xFF as arguments to C<vec>
-is deprecated. This will be a fatal error in Perl 5.32
+=item Use of strings with code points over 0xFF as arguments to vec is forbidden
 
-(D deprecated) You tried to use L<C<vec>|perlfunc/vec EXPR,OFFSET,BITS>
+(F) You tried to use L<C<vec>|perlfunc/vec EXPR,OFFSET,BITS>
 on a string containing a code point over 0xFF, which is nonsensical here.
 
-Such usage will be a fatal error in Perl 5.32.
+This became fatal in Perl 5.32.
 
 =item Use of tainted arguments in %s is deprecated
 
diff --git a/pod/perlfunc.pod b/pod/perlfunc.pod
index 196a88d71f..0e50132907 100644
--- a/pod/perlfunc.pod
+++ b/pod/perlfunc.pod
@@ -9669,10 +9669,8 @@ to try to write off the beginning of the string (i.e., negative OFFSET).
 If the string happens to be encoded as UTF-8 internally (and thus has
 the UTF8 flag set), L<C<vec>|/vec EXPR,OFFSET,BITS> tries to convert it
 to use a one-byte-per-character internal representation. However, if the
-string contains characters with values of 256 or higher, that conversion
-will fail, and a deprecation message will be raised.  In that situation,
-C<vec> will operate on the underlying buffer regardless, in its internal
-UTF-8 representation.  In Perl 5.32, this will be a fatal error.
+string contains characters with values of 256 or higher, a fatal error
+will occur.
 
 Strings created with L<C<vec>|/vec EXPR,OFFSET,BITS> can also be
 manipulated with the logical
diff --git a/t/lib/warnings/doop b/t/lib/warnings/doop
deleted file mode 100644
index 09db146737..0000000000
--- a/t/lib/warnings/doop
+++ /dev/null
@@ -1,14 +0,0 @@
-__END__
-# doop.c
-use utf8 ;
-$_ = "\x80  \xff" ;
-chop ;
-EXPECT
-########
-# NAME vec with above ff code points is deprecated
-my $foo = "\x{100}" . "\xff\xfe";
-eval { vec($foo, 1, 8) };
-no warnings 'deprecated';
-eval { vec($foo, 1, 8) };
-EXPECT
-Use of strings with code points over 0xFF as arguments to vec is deprecated. This will be a fatal error in Perl 5.32 at - line 2.
diff --git a/t/op/vec.t b/t/op/vec.t
index d3bb57dcd7..aa965785d3 100644
--- a/t/op/vec.t
+++ b/t/op/vec.t
@@ -10,6 +10,7 @@ use Config;
 
 plan(tests => 78);
 
+my $exception_134139 = "Use of strings with code points over 0xFF as arguments to vec is forbidden";
 
 is(vec($foo,0,1), 0);
 is(length($foo), undef);
@@ -65,18 +66,15 @@ $x = substr $foo, 1;
 is(vec($x, 0, 8), 255);
 $@ = undef;
 {
-    no warnings 'deprecated';
+    local $@;
     eval { vec($foo, 1, 8) };
-    ok(! $@);
+    like($@, qr/$exception_134139/,
+        "Caught exception: code point over 0xFF used as argument to vec");
     $@ = undef;
     eval { vec($foo, 1, 8) = 13 };
-    ok(! $@);
-    if ($::IS_EBCDIC) {
-        is($foo, "\x8c\x0d\xff\x8a\x69");
-    }
-    else {
-        is($foo, "\xc4\x0d\xc3\xbf\xc3\xbe");
-    }
+    #ok(! $@);
+    like($@, qr/$exception_134139/,
+        "Caught exception: code point over 0xFF used as argument to vec");
 }
 $foo = "\x{100}" . "\xff\xfe";
 $x = substr $foo, 1;
@@ -244,3 +242,13 @@ like($@, qr/^Modification of a read-only value attempted at /,
     $v = eval { RT131083(1, vec($s, $off, 8)); };
     like($@, qr/Out of memory!/, "RT131083 lval ~0");
 }
+
+{
+    # Adapting test formerly in t/lib/warnings/doop
+
+    local $@;
+    my $foo = "\x{100}" . "\xff\xfe";
+    eval { vec($foo, 1, 8) };
+    like($@, qr/$exception_134139/,
+        "RT 134139: Use of strings with code points over 0xFF as arguments to 'vec' is now forbidden");
+}
-- 
2.17.1


```
</details>

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @jkeenan

On Mon\, 27 May 2019 14&#8203;:08&#8203;:14 GMT\, jkeenan wrote&#8203;:
> On Sat\, 25 May 2019 02&#8203;:08&#8203;:59 GMT\, jkeenan@&#8203;pobox\.com wrote&#8203;:
> > pod/perldeprecation\.pod contains this entry for perl\-5\.32&#8203;:
> >
> > \#\#\#\#\#
> > Use of strings with code points over 0xFF as arguments to
> > "vec"
> >
> > "vec" views its string argument as a sequence of bits\. A
> > string containing a code point over 0xFF is nonsensical\.
> > This usage is deprecated in Perl 5\.28\, and will be removed
> > in Perl 5\.32\.
> > \#\#\#\#\#
> >
> > This entry was \(mostly\) added in this commit&#8203;:
> >
> > \#\#\#\#\#
> > commit 76aae383784f602979d5efc9c8828918d67827d8
> > Author&#8203;: Karl Williamson \<khw@&#8203;cpan\.org>
> > Date&#8203;:   Thu Jun 1 16&#8203;:10&#8203;:48 2017 \-0600
> >
> > perldeprecation&#8203;: Add vec\(\) with above\-FF code points
> > \#\#\#\#\#
> >
> > Make it so\.
> 
> Please review the patch attached\.  This is available for smoking in
> the following branch&#8203;:
> 
> smoke\-me/jkeenan/rt\-134139\-vec
> 
> Thank you very much\.

Reviewed by khw\.  Merged to blead with minor edits in commit da5a0da22ef5f5dd64e1fcdcac2c5ed1c0398085\.

Thank you very much\.
\-\- 
James E Keenan \(jkeenan@&#8203;cpan\.org\)

**Comment by p5pRT** [comment]
@jkeenan - Status changed from 'open' to 'pending release'
